### PR TITLE
REST API streaming support 

### LINKED ISSUE #495: Support state stream API
### Problem

The emulator does not currently implement the state stream API needed for events streaming

### Steps to Reproduce

Try using events streaming (i.e. `ws://localhost:8888/v1/subscribe_events`) and the endpoint will not be available because the state_stream.API instance used to instantiate the REST server is nil https://github.com/onflow/flow-emulator/blob/0cdc31ceecde40ee7ffbf70b8e77961b7bbd7ce4/server/access/rest.go#L99-L117

### Acceptance Criteria

Event streaming supported by emulator

### Context

Blocking adoption of event streaming API (i.e. https://github.com/onflow/fcl-js/issues/1791)

### PR DESCRIPTION
**Pull request by bluesign**
Closes #495 

## Description

Currently trie updates are not implemented but as-is it is enough to provide events, currently REST API does not allow state streaming AFAIK, so I marked as ready for review.  ( as it is a blocker for FCL work ) 

______

For contributor use:

- [X] Targeted PR against `master` branch
- [X] Linked to GitHub issue with discussion and accepted design OR link to spec that describes this work
- [X] Code follows the [standards mentioned here](https://github.com/onflow/flow-emulator/blob/master/CONTRIBUTING.md#styleguides)
- [ ] Updated relevant documentation
- [X] Re-reviewed `Files changed` in the GitHub PR explorer
- [ ] Added appropriate labels

### CODE REVIEW

**Comment by bluesign** [comment]
@peterargue can you have a look at this? I am not sure if we need execution streaming here or just events are enough

File: server/access/streamBackend.go [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+		getStartHeight:   getStartHeightFunc(blockchain),
+	}
+}
+
+var _ state_stream.API = &StateStreamBackend{}
+
+func getStartHeightFunc(blockchain *emulator.Blockchain) GetStartHeightFunc {
+	return func(blockID flow.Identifier, height uint64) (uint64, error) {
+		block, err := blockchain.GetBlockByID(blockID)
+		if err == nil {
```
Comment (sideninja) [inline server/access/streamBackend.go:70]: Maybe worth checking type of an error: 
```go
errors.Is(err, storage.ErrNotFound)
```

File: server/access/streamBackend.go [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+		}
+		return height, nil
+	}
+}
+
+func getExecutionDataFunc(blockchain *emulator.Blockchain) GetExecutionDataFunc {
+	return func(_ context.Context, height uint64) (*execution_data.BlockExecutionDataEntity, error) {
+		block, err := blockchain.GetBlockByHeight(height)
+
+		if err != nil {
```
Comment (sideninja) [inline server/access/streamBackend.go:81]: what if the returned error is not storage err not found? should you overwrite an error like so?

File: server/access/streamBackend.go [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+			return nil, storage.ErrNotFound
+		}
+		chunks := make([]*execution_data.ChunkExecutionData, 0)
+
+		for _, collectionGuarantee := range block.Payload.Guarantees {
```
Comment (sideninja) [inline server/access/streamBackend.go:86]: nit: move after the `blockchain.GetCollectionByID`

File: server/access/streamBackend.go [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+				if err != nil {
+					return nil, err
+				}
+				events = append(events, txResult.Events...)
+
+				var lightResult flow.LightTransactionResult
+				lightResult.TransactionID = txResult.TransactionID
+				lightResult.ComputationUsed = 0
+				lightResult.Failed = txResult.Status != 0
+				txResults = append(txResults, lightResult)
```
Comment (sideninja) [inline server/access/streamBackend.go:112]: nit change to inline:
```go
lightResult := flow.LightTransactionResult{
    TransactionID: txResult.TransactionID,
    Failed: txResult.ErrorMessage != "",
}
```
not so nit:
Failed should be determined by checking the presence of an error: `Failed: txResult.ErrorMessage != ""`.

File: server/access/streamBackend.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+			}
+
+			chunk := &execution_data.ChunkExecutionData{
+				Collection: collection,
+				Events:     events,
+				//TODO: add trie updates
+				TrieUpdate:         &ledger.TrieUpdate{},
```
Comment (sideninja) [inline server/access/streamBackend.go:119]: nit: remove this line, it's better to default to nil.

File: server/access/streamBackend.go [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+				TransactionResults: txResults,
+			}
+			chunks = append(chunks, chunk)
+		}
+
+		result := execution_data.NewBlockExecutionDataEntity(
+			flow.ZeroID,
```
Comment (sideninja) [inline server/access/streamBackend.go:126]: I believe you can calculate the ID like this:
```go
buf := new(bytes.Buffer)

serializer := execution_data.DefaultSerializer
err = serializer.Serialize(buf, executionData)
if err != nil {
    return nil, err
}

rootBlob := blobs.NewBlob(buf.Bytes())
rootID, err := flow.CidToId(rootBlob.Cid())
if err != nil {
    return nil, err
}

result := execution_data.NewBlockExecutionDataEntity(
    rootID,
    executionData,
)
```

File: server/access/streamBackend.go (hunk 1 above)
Comment (bluesign) [reply server/access/streamBackend.go]: is the logic here to try with blockID first then try with block height?

File: server/access/streamBackend.go (hunk 6 above)
Comment (bluesign) [reply server/access/streamBackend.go]: https://github.com/onflow/flow-emulator/pull/496/commits/0361e71aaf83d2e7c89c171301647cc8e4e775f8

File: server/access/streamBackend.go (hunk 5 above)
Comment (bluesign) [reply server/access/streamBackend.go]: https://github.com/onflow/flow-emulator/pull/496/commits/51b9ff62aa912fdb5288b65926db30c9266b592e

File: server/access/streamBackend.go (hunk 4 above)
Comment (bluesign) [reply server/access/streamBackend.go]: https://github.com/onflow/flow-emulator/pull/496/commits/46deb1901110a133cc1a5475b92ebf89706e2898

File: server/access/streamBackend.go (hunk 3 above)
Comment (bluesign) [reply server/access/streamBackend.go:109]: I didn't quite get this, you mean the  allocation ?

File: server/access/streamBackend.go (hunk 2 above)
Comment (bluesign) [reply server/access/streamBackend.go:100]: handled the not found case separately

File: server/access/streamBackend.go (hunk 1 above)
Comment (sideninja) [reply server/access/streamBackend.go]: I would say the logic is fine, but if GetBlockByID is not found then return the height, but if the GetBlockByID returns an unknown error then I believe we should return that error.

File: server/access/streamBackend.go (hunk 3 above)
Comment (sideninja) [reply server/access/streamBackend.go]: I mean the line 87 could be moved to line 92 after the call... super nit so ignore :D

**Review by sideninja — APPROVED** [review]

File: server/access/streamBackend.go (hunk 3 above)
Comment (bluesign) [reply server/access/streamBackend.go]: done :)

File: server/access/streamBackend.go (hunk 6 above)
Comment (bluesign) [reply server/access/streamBackend.go:158]: @sideninja this failed to provide ID unfortunately :(

**Comment by sideninja** [comment]
> > @bluesign good to merge?
> 
> Should be good, just this ID I will investigate a bit later but we can merge

yeah doesn't matter as much for now

**Pull request merged**